habitat/tasks/rearrange/rearrange_pick_task.py

```diff
@@ -0,0 +1,240 @@
+        timeout = 1000
+        for _attempt in range(timeout):
+            start_pos = orig_start_pos + np.random.normal(
+                0, self._config.BASE_NOISE, size=(3,)
+            )
+            targ_dist = np.linalg.norm((start_pos - orig_start_pos)[[0, 2]])
+
+            is_navigable = is_easy_init or sim._sim.pathfinder.is_navigable(
+                start_pos
+            )
+
+            if targ_dist > dist_thresh or not is_navigable:
+                continue
+
+            sim.set_state(state)
+
+            sim.set_robot_pos(start_pos[[0, 2]])
+
+            # Face the robot towards the object.
+            rel_targ = targ_pos - start_pos
+            angle_to_obj = get_angle(forward[[0, 2]], rel_targ[[0, 2]])
+            if np.cross(forward[[0, 2]], rel_targ[[0, 2]]) > 0:
+                angle_to_obj *= -1.0
+            sim.set_robot_rot(angle_to_obj)
+
+            # Make sure the robot is not colliding with anything in this
+            # position.
+            for _ in range(100):
+                sim.internal_step(-1)
+                colls = sim.get_collisions()
+                did_collide, details = rearrang_collision(
+                    colls,
+                    None,
+                    self._config.COUNT_OBJ_COLLISIONS,
+                    ignore_base=False,
+                )
+
+                if is_easy_init:
+                    # Only care about collisions between the robot and scene.
+                    did_collide = details.robo_scene_colls != 0
+
+                if did_collide:
+                    break
+
+            if not did_collide:
+                break
+
+        if _attempt == timeout - 1 and (not is_easy_init):
+            # print('Could not satisfy for %s! Trying with easy init' % sim.ep_info['scene_config_path'])
+            start_pos, angle_to_obj, sel_idx = self._gen_start_pos(sim, True)
+
+        sim.set_state(state)
+
+        return start_pos, angle_to_obj, sel_idx
+
+    def _should_prevent_grip(self, action_args):
+        return (
+            self._is_holding_obj()
+            and action_args.get("grip_ac", None) is not None
+            and action_args["grip_ac"] <= 0
+        )
+
+    def step(self, action, episode):
+        if "action_args" not in action or action["action_args"] is None:
+            action["action_args"] = {}
+        action_name = action["action"]
+        if isinstance(action_name, (int, np.integer)):
+            action_name = self.get_action_name(action_name)
+        assert (
+            action_name in self.actions
+        ), f"Can't find '{action_name}' action in {self.actions.keys()}."
+        action_args = action["action_args"]
+
+        if self._should_prevent_grip(action_args):
+            # No releasing the object once it is held.
+            action_args["grip_ac"] = None
+        obs = super().step(action=action, episode=episode)
+        obs = self._trans_obs(obs)
+
+        return obs
+
+    def _trans_obs(self, obs):
```

Comment:
This is bad code smell in general. ANd this is a pattern we should better extract.

Reply:
refactored and removed.